[MIGraphX EP] Add migx ep fp8 support and int4 weights

TedThemistokleous opened the pull request:
* Add fp8 and int4 types in supported list for Onnxruntime EP

* Add support for int4 inputs

Map things to int8 right now as we don't explicitly set an int4 input type and pack/unpack int4 operands

* Add flag to allow for fp8 quantization through Onnxruntime API

* Add fp8 quantization to the compile stage of the MIGraphX EP

Mirror the same calibration code we use for int8 and just change which quantize we call through the MIGraphx API

* cleanup logging

* Cleanup and encapsulate quantization / compile functions

- Add additional flags for fp8 thats shared for int8

- Add lockout warning message when int8/fp8 used at the same time

* Run lintrunner pass

* Fix session options inputs + add better logging.

Previous runs using session options failed as we were missing pulling in inputs from the python interface. This plus additional logging allowed me to track what options were invoked via env and what were added during the start of an inference session

* Fix naming for save/load path varibles to be consistent with  enable.

* Print only env variables that are set as warnings

need this so the user knows there's any of the environment variables running in the background to ensure proper consistently between runs.

---------

### Description
<!-- Describe your changes. -->
Changes to cleanup the MIGraphX EP quantization code as well as adding fp8 quantization support along with int4 support.

Cleanup changes handle a few instances of issues seen with the python interface when taking in provider options


### Motivation and Context
<!-- - Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here. -->

Required as we fix ignored flags when using provider_options for the MIGraphX EP
Adding fp8 quantization through the MIGraphX API
Adding int4 weight support for packed int4 weights for MIGraphX inference

tianleiwu commented on onnxruntime/python/onnxruntime_pybind_state.cc:
```diff
@@ -878,6 +879,16 @@ std::unique_ptr<IExecutionProvider> CreateExecutionProviderInstance(
+            params.migraphx_fp8_enable = false;
+          } else {
+            ORT_THROW(
+                "[ERROR] [MIGraphX] The value for the key 'migx_fp8_enable' should be"
```
migx_fp8_enable => migraphx_fp8_enable

TedThemistokleous replied:
Updated as well as cleaned up others

tianleiwu commented:
/azp run Windows ARM64 QNN CI Pipeline,Windows x64 QNN CI Pipeline,Windows CPU CI Pipeline,Windows GPU CUDA CI Pipeline,Windows GPU DML CI Pipeline,Windows GPU Doc Gen CI Pipeline,Windows GPU TensorRT CI Pipeline,ONNX Runtime Web CI Pipeline,Linux CPU CI Pipeline,Linux CPU Minimal Build E2E CI Pipeline

tianleiwu commented:
/azp run Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,onnxruntime-binary-size-checks-ci-pipeline,Big Models,Linux Android Emulator QNN CI Pipeline

tianleiwu commented:
/azp run Android CI Pipeline,iOS CI Pipeline,ONNX Runtime React Native CI Pipeline,CoreML CI Pipeline,Linux DNNL CI Pipeline,Linux MIGraphX CI Pipeline,Linux ROCm CI Pipeline

tianleiwu commented:
/azp run Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,onnxruntime-binary-size-checks-ci-pipeline,Big Models,Linux Android Emulator QNN CI Pipeline

tianleiwu commented:
/azp run Android CI Pipeline,iOS CI Pipeline,ONNX Runtime React Native CI Pipeline,CoreML CI Pipeline,Linux DNNL CI Pipeline,Linux MIGraphX CI Pipeline,Linux ROCm CI Pipeline

TedThemistokleous commented:
> @TedThemistokleous, please take a look at build error:
> 
> /onnxruntime_src/onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc:345:18: error: ‘migraphx_shape_fp8e4m3fn_type’ was not declared in this scope; did you mean ‘migraphx_shape_fp8e4m3fnuz_type’? 345 | mgx_type = migraphx_shape_fp8e4m3fn_type; | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~ | migraphx_shape_fp8e4m3fnuz_type /onnxruntime_src/onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc:348:18: error: ‘migraphx_shape_fp8e5m2_type’ was not declared in this scope; did you mean ‘migraphx_shape_uint32_type’? 348 | mgx_type = migraphx_shape_fp8e5m2_type; | ^~~~~~~~~~~~~~~~~~~~~~~~~~~ | migraphx_shape_uint32_type /onnxruntime_src/onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc:351:18: error: ‘migraphx_shape_fp8e5m2fnuz_type’ was not declared in this scope; did you mean ‘migraphx_shape_fp8e4m3fnuz_type’? 351 | mgx_type = migraphx_shape_fp8e5m2fnuz_type; | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ | migraphx_shape_fp8e4m3fnuz_type /onnxruntime_src/onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc: In function ‘void onnxruntime::calibrate_and_quantize(migraphx::api::program&, const migraphx::api::target&, migraphx::api::program_parameters, bool, bool, bool, bool, std::unordered_map<std::__cxx11::basic_string, float>&)’: /onnxruntime_src/onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc:1249:17: error: ‘quantize_fp8_options’ is not a member of ‘migraphx’ 1249 | migraphx::quantize_fp8_options quant_opts; | ^~~~~~~~~~~~~~~~~~~~ /onnxruntime_src/onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc:1250:7: error: ‘quant_opts’ was not declared in this scope 1250 | quant_opts.add_calibration_data(quant_params); | ^~~~~~~~~~ /onnxruntime_src/onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc:1251:17: error: ‘quantize_fp8’ is not a member of ‘migraphx’ 1251 | migraphx::quantize_fp8(prog, t, quant_opts);

Shoot, You'll need to use a later version of MIGraphx in your CI. We've added the quantize_fp8 function as part of our api recently. 

Where in your CI does it pull/use MIGraphX or do you pin things to a ROCm version?

tianleiwu commented:
> Shoot, You'll need to use a later version of MIGraphx in your CI. We've added the quantize_fp8 function as part of our api recently.
> 
> Where in your CI does it pull/use MIGraphX or do you pin things to a ROCm version?

This is the docker file used in CI:
https://github.com/microsoft/onnxruntime/blob/2d8b86e09a5fbc9aeec30069c9a1ad75325a5264/tools/ci_build/github/linux/docker/migraphx-ci-pipeline-env.Dockerfile

Related commit:
https://github.com/microsoft/onnxruntime/commit/28ee049e5698e8df2b7ee023f26487b99d273034

TedThemistokleous commented:
> > Shoot, You'll need to use a later version of MIGraphx in your CI. We've added the quantize_fp8 function as part of our api recently.
> > Where in your CI does it pull/use MIGraphX or do you pin things to a ROCm version?
> 
> This is the docker file used in CI: https://github.com/microsoft/onnxruntime/blob/2d8b86e09a5fbc9aeec30069c9a1ad75325a5264/tools/ci_build/github/linux/docker/migraphx-ci-pipeline-env.Dockerfile
> 
> Related commit: [28ee049](https://github.com/microsoft/onnxruntime/commit/28ee049e5698e8df2b7ee023f26487b99d273034)

Lets put this on ice then and I can gate this feature based on ROCm 6.4. That's where this change is targeted to for MIGraphX. In the meantime, do you want me to help updating your CI to ROCm 6.3? We just released ROCm 6.3.2 and it looks like you're using a ROCm 6.2.3 build in your CI instead

TedThemistokleous commented:
related: https://github.com/microsoft/onnxruntime/pull/23535

TedThemistokleous commented:
> > Shoot, You'll need to use a later version of MIGraphx in your CI. We've added the quantize_fp8 function as part of our api recently.
> > Where in your CI does it pull/use MIGraphX or do you pin things to a ROCm version?
> 
> This is the docker file used in CI: https://github.com/microsoft/onnxruntime/blob/2d8b86e09a5fbc9aeec30069c9a1ad75325a5264/tools/ci_build/github/linux/docker/migraphx-ci-pipeline-env.Dockerfile
> 
> Related commit: [28ee049](https://github.com/microsoft/onnxruntime/commit/28ee049e5698e8df2b7ee023f26487b99d273034)

Actually, ROCm 6.3.2 release is sufficient for this API call as we've added this in. Quick check shows this. CI change to ROCm 6.3.2 should be enough. 

```
~/AMDMIGraphX AMDMIGraphX ((rocm-6.3.2))git log | grep fp8
    fp8 functional not performance (#2845)
    use `optimize_module` for the fp8 JIT compilation  (#2656)
    * use eliminate_convert for the fp8
    Autocast_fp8 pass (#2527)
    * Autocast_fp8 pass
    * Address review: used contains to check fp8_types set
    Add `--fp8` option to quantize models in FP8 inside `migraphx-driver` (#2535)
    Enable simplify qdq to work with FP8 types, add warning for fp8 when loading mxr
    ad fp8 warnings (#2531)
    Update onnx proto and onnx parser to handle fp8 types (#2493)
    Only adding fp8e4m3fnuz in MIGraphX IR for now.
    fp16 and fp8 quantization to include subgraph and parameters
```

It would pull the existing package found on our public repo.radeon found here: https://repo.radeon.com/rocm/apt/6.3.2/pool/main/m/migraphx/

TedThemistokleous commented:
@tianleiwu let me know if you require anything else

TedThemistokleous commented:
Gating this behind ROCm 6.4 release changes.

tianleiwu commented:
/azp run Windows ARM64 QNN CI Pipeline,Windows x64 QNN CI Pipeline,Windows CPU CI Pipeline,Windows GPU CUDA CI Pipeline,Windows GPU DML CI Pipeline,Windows GPU Doc Gen CI Pipeline,Windows GPU TensorRT CI Pipeline,ONNX Runtime Web CI Pipeline,Linux CPU CI Pipeline,Linux CPU Minimal Build E2E CI Pipeline

tianleiwu commented:
/azp run Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,onnxruntime-binary-size-checks-ci-pipeline,Big Models,Linux Android Emulator QNN CI Pipeline

tianleiwu commented:
/azp run Android CI Pipeline,iOS CI Pipeline,ONNX Runtime React Native CI Pipeline,CoreML CI Pipeline,Linux DNNL CI Pipeline,Linux MIGraphX CI Pipeline,Linux ROCm CI Pipeline

tianleiwu commented:
/azp run Win_TRT_Minimal_CUDA_Test_CI

tianleiwu commented:
@TedThemistokleous, please take a look at build error:
```
/onnxruntime_src/onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc: In member function ‘void onnxruntime::MIGraphXExecutionProvider::get_flags_from_env()’:
/onnxruntime_src/onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc:185:5: error: ‘fp8_enable’ was not declared in this scope; did you mean ‘fp8_enable_’?
  185 |     fp8_enable = false;
      |     ^~~~~~~~~~
      |     fp8_enable_
/onnxruntime_src/onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc: In function ‘bool onnxruntime::getMIGraphXType(ONNXTensorElementDataType, migraphx_shape_datatype_t&)’:
/onnxruntime_src/onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc:356:18: error: ‘migraphx_shape_fp8e4m3fn_type’ was not declared in this scope; did you mean ‘migraphx_shape_fp8e4m3fnuz_type’?
  356 |       mgx_type = migraphx_shape_fp8e4m3fn_type;
      |                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                  migraphx_shape_fp8e4m3fnuz_type
/onnxruntime_src/onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc:359:18: error: ‘migraphx_shape_fp8e5m2_type’ was not declared in this scope; did you mean ‘migraphx_shape_uint32_type’?
  359 |       mgx_type = migraphx_shape_fp8e5m2_type;
      |                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                  migraphx_shape_uint32_type
/onnxruntime_src/onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc:362:18: error: ‘migraphx_shape_fp8e5m2fnuz_type’ was not declared in this scope; did you mean ‘migraphx_shape_fp8e4m3fnuz_type’?
  362 |       mgx_type = migraphx_shape_fp8e5m2fnuz_type;
      |                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                  migraphx_shape_fp8e4m3fnuz_type
```

TedThemistokleous commented:
> @TedThemistokleous, please take a look at build error:
> 
> ```
> /onnxruntime_src/onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc: In member function ‘void onnxruntime::MIGraphXExecutionProvider::get_flags_from_env()’:
> /onnxruntime_src/onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc:185:5: error: ‘fp8_enable’ was not declared in this scope; did you mean ‘fp8_enable_’?
>   185 |     fp8_enable = false;
>       |     ^~~~~~~~~~
>       |     fp8_enable_
> /onnxruntime_src/onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc: In function ‘bool onnxruntime::getMIGraphXType(ONNXTensorElementDataType, migraphx_shape_datatype_t&)’:
> /onnxruntime_src/onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc:356:18: error: ‘migraphx_shape_fp8e4m3fn_type’ was not declared in this scope; did you mean ‘migraphx_shape_fp8e4m3fnuz_type’?
>   356 |       mgx_type = migraphx_shape_fp8e4m3fn_type;
>       |                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>       |                  migraphx_shape_fp8e4m3fnuz_type
> /onnxruntime_src/onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc:359:18: error: ‘migraphx_shape_fp8e5m2_type’ was not declared in this scope; did you mean ‘migraphx_shape_uint32_type’?
>   359 |       mgx_type = migraphx_shape_fp8e5m2_type;
>       |                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~
>       |                  migraphx_shape_uint32_type
> /onnxruntime_src/onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc:362:18: error: ‘migraphx_shape_fp8e5m2fnuz_type’ was not declared in this scope; did you mean ‘migraphx_shape_fp8e4m3fnuz_type’?
>   362 |       mgx_type = migraphx_shape_fp8e5m2fnuz_type;
>       |                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>       |                  migraphx_shape_fp8e4m3fnuz_type
> ```

You need to go to a newer ROCm version in your CI. FP8 datatypes were added in the previous new year.

As a side note, how often does your team update the baseline ROCm release?

tianleiwu commented:
@TedThemistokleous, could you add the change of MigraphX pipeline of https://github.com/microsoft/onnxruntime/pull/23535 to this PR?  Do not change ROCm CI pipeline.

TedThemistokleous commented:
@tianleiwu  did these pipelines end up getting updated?

tianleiwu commented:
@TedThemistokleous, I think Migraphx pipeline has not  updated: https://github.com/microsoft/onnxruntime/pull/23535 is not merged.

TedThemistokleous commented:
> @TedThemistokleous, I think Migraphx pipeline has not updated: #23535 is not merged.

Merged main into #23535 this and awaiting review.

tianleiwu commented:
/azp run Linux QNN CI Pipeline, Win_TRT_Minimal_CUDA_Test_CI, Windows ARM64 QNN CI Pipeline, Windows GPU Doc Gen CI Pipeline, Windows x64 QNN CI Pipeline

tianleiwu commented:
I've verified that the build of Linux MIGraphX CI Pipeline is good with this PR.

TedThemistokleous commented:
Merged the other PR to update your CI, is there anything else required here?

tianleiwu commented:
@TedThemistokleous, could you merge latest main branch? Some required CI pipelines are not in your branch.

TedThemistokleous commented:
@tianleiwu  done merged main.

tianleiwu commented:
/azp run Linux QNN CI Pipeline, Win_TRT_Minimal_CUDA_Test_CI, Windows ARM64 QNN CI Pipeline, Windows GPU Doc Gen CI Pipeline, Windows x64 QNN CI Pipeline
